common/unsafe/src/main/java/org/apache/spark/unsafe/memory/UnsafeMemoryAllocator.java

```diff
@@ -36,22 +42,34 @@ public MemoryBlock allocate(long size) throws OutOfMemoryError {
 
   @Override
   public void free(MemoryBlock memory) {
-    assert (memory.obj == null) :
+    if (memory == OffHeapMemoryBlock.NULL) return;
+    assert(memory instanceof OffHeapMemoryBlock);
+    assert (memory.getBaseObject() == null) :
       "baseObject not null; are you trying to use the off-heap allocator to free on-heap memory?";
-    assert (memory.pageNumber != MemoryBlock.FREED_IN_ALLOCATOR_PAGE_NUMBER) :
+    assert (memory.getPageNumber() != MemoryBlock.FREED_IN_ALLOCATOR_PAGE_NUMBER) :
       "page has already been freed";
-    assert ((memory.pageNumber == MemoryBlock.NO_PAGE_NUMBER)
-            || (memory.pageNumber == MemoryBlock.FREED_IN_TMM_PAGE_NUMBER)) :
+    assert ((memory.getPageNumber() == MemoryBlock.NO_PAGE_NUMBER)
+            || (memory.getPageNumber() == MemoryBlock.FREED_IN_TMM_PAGE_NUMBER)) :
       "TMM-allocated pages must be freed via TMM.freePage(), not directly in allocator free()";
 
     if (MemoryAllocator.MEMORY_DEBUG_FILL_ENABLED) {
       memory.fill(MemoryAllocator.MEMORY_DEBUG_FILL_FREED_VALUE);
     }
+
     Platform.freeMemory(memory.offset);
+
     // As an additional layer of defense against use-after-free bugs, we mutate the
     // MemoryBlock to reset its pointer.
-    memory.offset = 0;
+    memory.resetObjAndOffset();
     // Mark the page as freed (so we can detect double-frees).
-    memory.pageNumber = MemoryBlock.FREED_IN_ALLOCATOR_PAGE_NUMBER;
+    memory.setPageNumber(MemoryBlock.FREED_IN_ALLOCATOR_PAGE_NUMBER);
+  }
+
+  public OffHeapMemoryBlock reallocate(OffHeapMemoryBlock block, long oldSize, long newSize) {
+    OffHeapMemoryBlock mb = this.allocate(newSize);
+    if (block.getBaseOffset() != 0)
+      MemoryBlock.copyMemory(block, mb, oldSize);
```

Comment:
`block` and `mb` point to the same data, is this operation safe?

Reply:
Good point. At least, we can skip copy.

Reply:
I hope mb points to the new area. But, let me avoid free if they are the same.

Reply:
ah sorry I misread the code, I thought you were calling `MemoryBlock#allocate`. I think we don't need the extra check below.